clients/python-wrapper/lakefs/object.py

```diff
@@ -357,62 +307,96 @@ def _get_range_string(start, read_bytes=None):
             raise ValueError(f"invalid write mode: '{mode}'. WriteModes: {WriteModes}")
 
-        if 'x' in mode and self.exists():  # Requires explicit create
-            raise ObjectExistsException
+        self.content_type = content_type
+        self.metadata = metadata
 
-        binary_mode = 'b' in mode
-        if binary_mode and isinstance(data, str):
-            content = data.encode('utf-8')
-        elif not binary_mode and isinstance(data, bytes):
-            content = data.decode('utf-8')
+        open_kwargs = {
+            "encoding": "utf-8" if 'b' not in mode else None,
+            "mode": 'wb+' if 'b' in mode else 'w+',
+            "buffering": _BUFFER_SIZE,
+            "max_size": _BUFFER_SIZE,
+        }
+        self._fd = tempfile.SpooledTemporaryFile(**open_kwargs)  # pylint: disable=consider-using-with
+        super().__init__(obj, mode, pre_sign, client)
+
+    @property
+    def pre_sign(self) -> bool:
+        """
+        Returns whether the pre_sign mode is enabled
+        """
+        if self._pre_sign is None:
+            self._pre_sign = self._client.storage_config.pre_sign_support
+        return self._pre_sign
+
+    @pre_sign.setter
+    def pre_sign(self, value: bool) -> None:
+        """
+        Set the pre_sign mode to value
+        :param value: The new value for pre_sign mode
+        """
+        self._pre_sign = value
+
+    def flush(self) -> None:
+        """
+        Flush buffer to file. Prevent flush if total write size is still smaller than _BUFFER_SIZE so that we avoid
+        unnecessary write to disk.
+        """
+        # Don't flush buffer to file if we didn't exceed buffer size
+        # We want to avoid using the file if possible
+        if self._pos > _BUFFER_SIZE:
+            self._fd.flush()
+
+    def write(self, s: AnyStr) -> int:
+        """
+        Write data to buffer
+        :param s: The data to write
+        :return: The number of bytes written to buffer
+        """
+        binary_mode = 'b' in self._mode
+        if binary_mode and isinstance(s, str):
+            contents = s.encode('utf-8')
+        elif not binary_mode and isinstance(s, bytes):
+            contents = s.decode('utf-8')
         else:
-            content = data
+            contents = s
 
-        is_presign = pre_sign if pre_sign is not None else self._client.storage_config.pre_sign_support
-        with api_exception_handler(_io_exception_handler):
-            stats = self._upload_presign(content, content_type, metadata) if is_presign \
-                else self._upload_raw(content, content_type, metadata)
-            self._stats = ObjectStats(**stats.dict())
+        count = self._fd.write(contents)
+        self._pos += count
+        return count
 
-        return self
+    def close(self) -> None:
+        """
+        Write the data to the lakeFS server and close open descriptors
+        """
+        stats = self._upload_presign() if self.pre_sign else self._upload_raw()
```

Comment:
since all writes are gone unless this successfully happens, maybe raise [UserWarning](https://docs.python.org/3/library/exceptions.html#UserWarning) if writer is destructed after being written to, but `close()` not called..?

Reply:
Opened: https://github.com/treeverse/lakeFS/issues/7047